File: Core/src/com/stusherwin/midiapp/core/midi/modules/MidiInput.java
package com.stusherwin.midiapp.core.midi.modules;

import com.stusherwin.midiapp.core.InitializationException;
import com.stusherwin.midiapp.core.modules.SourceModule;
import com.stusherwin.midiapp.core.midi.MidiDevice;
import com.stusherwin.midiapp.core.midi.MidiDeviceInputListener;
import com.stusherwin.midiapp.core.midi.MidiEvent;

public class MidiInput extends SourceModule<MidiEvent> implements MidiDeviceInputListener {
    private MidiDevice midiDevice;

    public MidiInput(MidiDevice midiDevice) {
        this.midiDevice = midiDevice;

        this.midiDevice.setInputListener(this);
    }

    @Override
    public void init() throws InitializationException {
        this.midiDevice.init();
    }

    @Override
    public void destroy() {
        this.midiDevice.destroy();
    }

    @Override
    public void onMidiEvent(MidiEvent midiEvent) {
        this.output().observer().onNext(midiEvent);
    }
}

File: Core/src/com/stusherwin/midiapp/core/midi/modules/MidiInput.scala
package com.stusherwin.midiapp.core.midi.modules

import com.stusherwin.midiapp.core.modules.SourceModule
import com.stusherwin.midiapp.core.midi.MidiDevice
import com.stusherwin.midiapp.core.midi.MidiDeviceInputListener
import com.stusherwin.midiapp.core.midi.MidiEvent

class MidiInput(midiDevice : MidiDevice) extends SourceModule[MidiEvent] with MidiDeviceInputListener {
  midiDevice.setInputListener(this)

    override def init() : Unit = {
      midiDevice.init()
    }

    override def destroy() : Unit = {
      midiDevice.destroy()
    }

    override def onMidiEvent(midiEvent : MidiEvent) : Unit = {
      output.observer.onNext(midiEvent)
    }
}